internal/datacoord/services.go

```diff
@@ -160,39 +171,90 @@ func (s *Server) Flush(ctx context.Context, req *datapb.FlushRequest) (*datapb.F
 			return nil
 		}, retry.Attempts(60)) // about 3min
 		if err != nil {
-			return &datapb.FlushResponse{
-				Status: merr.Status(err),
-			}, nil
+			return nil, err
 		}
 
 		if isUnimplemented {
 			// For compatible with rolling upgrade from version 2.2.x,
 			// fall back to the flush logic of version 2.2.x;
 			log.Warn("DataNode FlushChannels unimplemented", zap.Error(err))
-			ts = 0
+			flushTs = 0
 		}
 	}
 
 	log.Info("flush response with segments",
-		zap.Int64("collectionID", req.GetCollectionID()),
+		zap.Int64("collectionID", collectionID),
 		zap.Int64s("sealSegments", lo.Keys(sealedSegmentsIDDict)),
 		zap.Int("flushedSegmentsCount", len(flushSegmentIDs)),
 		zap.Time("timeOfSeal", timeOfSeal),
-		zap.Uint64("flushTs", ts),
-		zap.Time("flushTs in time", tsoutil.PhysicalTime(ts)))
+		zap.Uint64("flushTs", flushTs),
+		zap.Time("flushTs in time", tsoutil.PhysicalTime(flushTs)))
 
-	return &datapb.FlushResponse{
-		Status:          merr.Success(),
-		DbID:            req.GetDbID(),
-		CollectionID:    req.GetCollectionID(),
+	return &datapb.FlushResult{
+		CollectionID:    collectionID,
 		SegmentIDs:      lo.Keys(sealedSegmentsIDDict),
 		TimeOfSeal:      timeOfSeal.Unix(),
 		FlushSegmentIDs: flushSegmentIDs,
-		FlushTs:         ts,
+		FlushTs:         flushTs,
 		ChannelCps:      channelCPs,
 	}, nil
 }
 
+func (s *Server) FlushAll(ctx context.Context, req *datapb.FlushAllRequest) (*datapb.FlushAllResponse, error) {
+	log := log.Ctx(ctx)
+	log.Info("receive flushAll request")
+	ctx, sp := otel.Tracer(typeutil.DataCoordRole).Start(ctx, "DataCoord-Flush")
+	defer sp.End()
+
+	if err := merr.CheckHealthy(s.GetStateCode()); err != nil {
+		log.Info("server is not healthy", zap.Error(err), zap.Any("stateCode", s.GetStateCode()))
+		return &datapb.FlushAllResponse{
+			Status: merr.Status(err),
+		}, nil
+	}
+
+	resp, err := s.broker.ShowCollectionIDs(ctx, req.GetDbName())
+	if err != nil {
+		return &datapb.FlushAllResponse{
+			Status: merr.Status(err),
+		}, nil
+	}
+
+	// generate a timestamp timeOfSeal, all data before timeOfSeal is guaranteed to be sealed or flushed
+	ts, err := s.allocator.AllocTimestamp(ctx)
+	if err != nil {
+		log.Warn("unable to alloc timestamp", zap.Error(err))
+		return nil, err
+	}
+
+	dbCollections := resp.GetDbCollections()
+	wg := errgroup.Group{}
```

Comment:
add go routine limit
like
wg.SetLimit(100) so there won't be too many goroutines at same time